src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -639,3 +640,151 @@ int32_t CryptoNative_SslGetCurrentCipherId(SSL* ssl, int32_t* cipherId)
 
     return 1;
 }
+
+// This function generates key pair and creates simple certificate.
+static int MakeSelfSignedCertificate(X509 * cert, EVP_PKEY* evp)
+{
+    RSA* rsa = CryptoNative_RsaCreate();
+    ASN1_TIME* time = ASN1_TIME_new();
+    BIGNUM* bn = BN_new();
+    BN_set_word(bn, RSA_F4);
+    X509_NAME * asnName;
+    unsigned char * name = (unsigned char*)"localhost";
+
+    int ret = 0;
+
+    if (rsa != NULL && CryptoNative_RsaGenerateKeyEx(rsa, 2048, bn) == 1)
+    {
+        if (CryptoNative_EvpPkeySetRsa(evp, rsa) == 1)
+        {
+            rsa = NULL;
+        }
+
+        X509_set_pubkey(cert, evp);
+
+        asnName = X509_get_subject_name(cert);
+        X509_NAME_add_entry_by_txt(asnName, "CN", MBSTRING_ASC, name, -1, -1, 0);
+
+        asnName =  X509_get_issuer_name(cert);
+        X509_NAME_add_entry_by_txt(asnName, "CN", MBSTRING_ASC, name, -1, -1, 0);
+
+        ASN1_TIME_set(time, 0);
+        X509_set1_notBefore(cert, time);
+        X509_set1_notAfter(cert, time);
+
+        ret = X509_sign(cert, evp, EVP_sha256());
+    }
+
+    if (bn != NULL)
+    {
+        BN_free(bn);
+    }
+
+    if (rsa != NULL)
+    {
+        RSA_free(rsa);
+    }
+
+    if (time != NULL)
+    {
+        ASN1_TIME_free(time);
+    }
+
+    return ret;
+}
+
+int32_t CryptoNative_OpenSslGetProtocolSupport(SslProtocols protocol)
```

Comment:
This is a lot of native code to add for a test probe, especially with needing to add creating a self-signed cert.  If it's just for tests can we detect the state somehow else?  Maybe move the cert generation to managed code and leave this as passing in a cert and a protocol? (Then it's easier to expand to test for the combinatorics of {RSA,DSA,ECDSA} x { TLS protocol versions })

Reply:
that is where I started. But the PlatfdormDetection build did not work for me the way other tests projects work and I was getting unresolved references. I can try to ping some build experts.

Reply:
While I'd prefer to see this as

```suggestion
int32_t CryptoNative_OpenSslGetProtocolSupport(SslProtocols protocol, X509* cert, EVP_PKEY* key)
```

and the cert generation moved to managed code (and then no longer need the X509 mutation functions), if this is what we have to do based on the calendar, it's OK for a test probe.